### PR TITLE
Performance optimizations (without regexp / fmt)

### PR DESCRIPTION
**Pull request by im-kulikov**

### CODE REVIEW

**Comment by emalm** [comment]
Thanks for the contribution, @im-kulikov ! We'll prioritize this for the CF Diego team to evaluate soon.

Out of curiosity, do you have a sense of how much these changes improve performance in particular situations? Would it be worth adding a benchmark test to ensure they do not regress?

Thanks again,
Eric, CF Diego PM

**Comment by im-kulikov** [comment]
kk

**Comment by im-kulikov** [comment]
```
name        old time/op    new time/op    delta
ToBytes-8      569ns ± 5%     148ns ± 3%  -73.97%  (p=0.000 n=9+10)
ByteSize-8     441ns ± 5%     276ns ± 4%  -37.41%  (p=0.000 n=10+9)

name        old alloc/op   new alloc/op   delta
ToBytes-8       118B ± 0%       16B ± 0%  -86.44%  (p=0.000 n=10+10)
ByteSize-8     44.0B ± 0%     40.0B ± 0%   -9.09%  (p=0.000 n=10+10)

name        old allocs/op  new allocs/op  delta
ToBytes-8       4.00 ± 0%      2.00 ± 0%  -50.00%  (p=0.000 n=10+10)
ByteSize-8      5.00 ± 0%      3.00 ± 0%  -40.00%  (p=0.000 n=10+10)
```

**Comment by im-kulikov** [comment]
is there any news?

File: bytes.go [hunk 1]
```diff
@@ -77,30 +73,30 @@ func ToMegabytes(s string) (uint64, error) {
...
 // TB = T = TiB = 1024 * G
 func ToBytes(s string) (uint64, error) {
-	parts := bytesPattern.FindStringSubmatch(strings.TrimSpace(s))
-	if len(parts) < 3 {
-		return 0, invalidByteQuantityError
-	}
+	s = strings.TrimSpace(s)
+	s = strings.ToUpper(s)
+	for i, v := range s {
+		if unicode.IsLetter(v) {
```
Comment (jvshahid) [inline bytes.go]: Can we use [IndexFunc](https://golang.org/pkg/strings/#IndexFunc) instead of the for loop ?

Comment (im-kulikov) [reply bytes.go]: hm.. ok.. but, what the reason?
```go
// IndexFunc returns the index into s of the first Unicode
// code point satisfying f(c), or -1 if none do.
func IndexFunc(s string, f func(rune) bool) int {
	return indexFunc(s, f, true)
}

// indexFunc is the same as IndexFunc except that if
// truth==false, the sense of the predicate function is
// inverted.
func indexFunc(s string, f func(rune) bool, truth bool) int {
	for i, r := range s {
		if f(r) == truth {
			return i
		}
	}
	return -1
}
```

Comment (im-kulikov) [reply bytes.go]: two additional calls

**Comment by im-kulikov** [comment]
```
name        old time/op    new time/op    delta
ToBytes-8      569ns ± 5%     164ns ± 2%  -71.15%  (p=0.000 n=9+10)
ByteSize-8     441ns ± 5%     276ns ± 2%  -37.34%  (p=0.000 n=10+10)

name        old alloc/op   new alloc/op   delta
ToBytes-8       118B ± 0%       16B ± 0%  -86.44%  (p=0.000 n=10+10)
ByteSize-8     44.0B ± 0%     40.0B ± 0%   -9.09%  (p=0.000 n=10+10)

name        old allocs/op  new allocs/op  delta
ToBytes-8       4.00 ± 0%      2.00 ± 0%  -50.00%  (p=0.000 n=10+10)
ByteSize-8      5.00 ± 0%      3.00 ± 0%  -40.00%  (p=0.000 n=10+10)
```

File: bytes.go [hunk 2]
```diff
@@ -77,30 +73,33 @@ func ToMegabytes(s string) (uint64, error) {
...
-	}
+	s = strings.TrimSpace(s)
+	s = strings.ToUpper(s)
 
-	value, err := strconv.ParseFloat(parts[1], 64)
-	if err != nil || value <= 0 {
-		return 0, invalidByteQuantityError
-	}
+	i := strings.IndexFunc(s, func(v rune) bool {
+		return unicode.IsLetter(v)
```
Comment (jvshahid) [inline bytes.go]: I think you can just pass `unicode.IsLetter` instead of wrapping it in a function.

File: bytes.go [hunk 3]
```diff
@@ -77,30 +73,33 @@ func ToMegabytes(s string) (uint64, error) {
...
-		bytes = uint64(value * TERABYTE)
-	case "G":
-		bytes = uint64(value * GIGABYTE)
-	case "M":
-		bytes = uint64(value * MEGABYTE)
-	case "K":
-		bytes = uint64(value * KILOBYTE)
-	case "B":
-		bytes = uint64(value * BYTE)
+	if i != -1 {
```
Comment (jvshahid) [inline bytes.go]: can you flip this test. I personally prefer 
```
if i == -1 {
  return error
}
continue processing
```

over 

```
if i != -1 {
  continue processing
}
return error
```

Comment (im-kulikov) [reply bytes.go]: this will complicate the code for understanding, I think, and add code dubbing

Comment (im-kulikov) [reply bytes.go]: Before:

```go
func ToBytes(s string) (uint64, error) {
	s = strings.TrimSpace(s)
	s = strings.ToUpper(s)

	i := strings.IndexFunc(s, unicode.IsLetter)

	if i != -1 {
		bytesString, multiple := s[:i], s[i:]
		bytes, err := strconv.ParseFloat(bytesString, 64)
		if err != nil || bytes <= 0 {
			return 0, invalidByteQuantityError
		}

		switch multiple {
		case "T", "TB", "TIB":
			return uint64(bytes * TERABYTE), nil
		case "G", "GB", "GIB":
			return uint64(bytes * GIGABYTE), nil
		case "M", "MB", "MIB":
			return uint64(bytes * MEGABYTE), nil
		case "K", "KB", "KIB":
			return uint64(bytes * KILOBYTE), nil
		case "B":
			return uint64(bytes), nil
		}
	}

	return 0, invalidByteQuantityError
}
```

After:
```go
func ToBytes(s string) (uint64, error) {
	s = strings.TrimSpace(s)
	s = strings.ToUpper(s)

	i := strings.IndexFunc(s, unicode.IsLetter)

	if i == -1 {
		return 0, invalidByteQuantityError
	}

	bytesString, multiple := s[:i], s[i:]
	bytes, err := strconv.ParseFloat(bytesString, 64)
	if err != nil || bytes <= 0 {
		return 0, invalidByteQuantityError
	}

	switch multiple {
	case "T", "TB", "TIB":
		return uint64(bytes * TERABYTE), nil
	case "G", "GB", "GIB":
		return uint64(bytes * GIGABYTE), nil
	case "M", "MB", "MIB":
		return uint64(bytes * MEGABYTE), nil
	case "K", "KB", "KIB":
		return uint64(bytes * KILOBYTE), nil
	case "B":
		return uint64(bytes), nil
	default:
		return 0, invalidByteQuantityError
	}
}
```

What you think?

File: bytes.go (hunk 2 above)
Comment (im-kulikov) [reply bytes.go]: yep, you're right

File: bytes.go (hunk 3 above)
Comment (jvshahid) [reply bytes.go]: I think that's fine. I prefer reducing nesting and the duplication isn't any worse than it used to be before.

Comment (im-kulikov) [reply bytes.go]: kk

Comment (im-kulikov) [reply bytes.go]: fixed

**Pull request closed without merging**